aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -790,36 +804,151 @@ Tensor select_sparse_csr(const Tensor& self, int64_t dim, int64_t index) {
-    // current version is more readable and easier to understand.
-    return self.to_sparse().select(dim, index);
+        select_name, ": selecting sparse dimensions is not implemented for batched sparse compressed tensors.")
+    TORCH_INTERNAL_ASSERT(dim == 0 || dim == 1);
+
+    DimVector blocksize{1, 1};
+    AT_DISPATCH_PLAIN_SPARSE_COMPRESSED_LAYOUTS(self.layout(), "select", [&] {}, [&] {
+      blocksize[0] = std::max<int64_t>(1, self.values().size(n_batch + 1));
+      blocksize[1] = std::max<int64_t>(1, self.values().size(n_batch + 2));
+    });
+
+    DimVector output_shape = DimVector(self.sizes().slice(1, self.dim()-1));
+    if (dim == 1) {
+      output_shape[0] = self.size(0);
+    }
+
+    auto indices_options = compressed_indices.options();
+    int64_t fast_dim = AT_DISPATCH_ROW_SPARSE_COMPRESSED_LAYOUTS(self.layout(), "select", [&]() { return 0; }, [&]() { return 1; });
+    int64_t start = -1;
+    int64_t end = -1;
+    int64_t select_nnz{};
+    Tensor dim_indices;
+    Tensor indices;
+    Tensor values;
+    bool is_copy = false;
+    if (dim == fast_dim) {
+      // select is always a view operation
+      Tensor start_end = compressed_indices.narrow(0, index / blocksize[0], 2).cpu();
+      AT_DISPATCH_INDEX_TYPES(start_end.scalar_type(), "select", [&]() {
+        auto accessor = start_end.accessor<index_t, 1>();
+        start = accessor[0];
+        end = accessor[1];
+      });
+      indices = plain_indices.slice(0, start, end);
+      values = self.values().slice(0, start, end);
+      select_nnz = end - start;
+    } else {
+      dim_indices = at::where(plain_indices.eq(index / blocksize[0]))[0];
+      // Assume that dim_indices is a sorted sequence of non-negative
+      // distinct integers. Below we'll try to solve `dim_indices ==
+      // arange(start, stop, step)`. If the solution exists then the
+      // select will be a view operation also for the `dim !=
+      // fast_dim` case.
+      select_nnz = dim_indices.numel();
+      int64_t step = -1;
+      if (select_nnz == 0) {
+        // trivial solution
+        start = end = 0;
+        step = 1;
+      } else {
+        Tensor dim_indices_cpu = dim_indices.cpu();
+        AT_DISPATCH_INDEX_TYPES(dim_indices_cpu.scalar_type(), "select", [&]() {
+          auto accessor = dim_indices_cpu.accessor<index_t, 1>();
+          if (accessor[select_nnz-1] - accessor[0] == select_nnz - 1) {
+            // special solution, avoids scanning dim_indices content
+            start = accessor[0];
+            end = start + select_nnz;
+            step = 1;
+          } else if (select_nnz > 1) {
+            // general solution, must scan the dim_indices content
+            int64_t step_candidate = accessor[1] - accessor[0];
+            for (int i=2; i<select_nnz; i++) {
+              if (accessor[i] - accessor[i-1] != step_candidate) {
+                step_candidate = 0;  // no solution
+                break;
+              }
+            }
```

Comment:
Can we simplify it with something like
```python
possible_steps = dim_indices[1:] - dim_indices[:-1]
solution_exists = (possible_steps == possible_steps[0]).all().item()
```

Reply:
Using now:
```c++
            Tensor possible_steps = dim_indices_cpu.slice(0, 1).sub(dim_indices_cpu.slice(0, 0, -1));
            index_t min_step = possible_steps.min().item<index_t>();
            index_t max_step = possible_steps.max().item<index_t>();
            if (min_step == max_step) {
              // found a general solution:
              step = max_step;
              start = accessor[0];
              end = start + select_nnz * step;
            }
```

Reply:
I am fine with that but do speculate that `all` might be short-circuited.

Reply:
The variant
```
possible_steps = dim_indices[1:] - dim_indices[:-1]
solution_exists = (possible_steps == possible_steps[0]).all().item()
```
has two full passes and one possibly short-circuited pass over data:
- full pass 1: `dim_indices[1:] - dim_indices[:-1]`
- full pass 2: `possible_steps == possible_steps[0]`
- possibly short-circuited pass: `(...).all()`

The current variant
```
Tensor possible_steps = dim_indices_cpu.slice(0, 1).sub(dim_indices_cpu.slice(0, 0, -1));
            index_t min_step = possible_steps.min().item<index_t>();
            index_t max_step = possible_steps.max().item<index_t>();
```
has three full passes: `sub`, `min`, and `max` method calls.

The original variant
```
           int64_t step_candidate = accessor[1] - accessor[0];
            for (int i=2; i<select_nnz; i++) {
              if (accessor[i] - accessor[i-1] != step_candidate) {
                step_candidate = 0;  // no solution
                break;
              }
            }
```
has zero full passes and one possibly short-circuited pass but it has a sequential loop.

Reply:
Here's another variant:
```c++
            index_t step_candidate = accessor[1] - accessor[0];
            Tensor tmp = at::arange(accessor[0], accessor[0] + select_nnz * step_candidate, step_candidate, indices_options);
            bool has_solution = (tmp.eq(dim_indices)).all().item<bool>();
```
@nikitaved any preferences?

Reply:
Well, maybe we can make it really a single pass by convolving with a kernel `[-1, 1]`...

Reply:
Convolve is a linear operation and I don't see how it would give the correct solution in general.

Reply:
Here's one more variant:
```c++
        Tensor first_second = dim_indices.narrow(0, 0, 2).cpu();
        AT_DISPATCH_INDEX_TYPES(first_second.scalar_type(), "select", [&]() {
          auto accessor = first_second.accessor<index_t, 1>();
          start = accessor[0];
          step = accessor[1] - start;
          end = start + select_nnz * step;
        });
        Tensor tmp = at::arange(start, end, step, indices_options);
        if (!(tmp.eq(dim_indices).all().item<bool>())) {
          // no solution
          start = end = step = -1;
        }
```

Reply:
> Convolve is a linear operation and I don't see how it would give the correct solution in general.

We are computing `arr[1:] - art[:-1]`. How is that not a linear operation and/or a convolution?

Reply:
Closing as out of date. See the new code for details